### PR TITLE
CIRC-1223 fix fetching accounts for refund

### PR DESCRIPTION
**Pull request by roman-barannyk**
## Purpose
Update logic for refund renew according to the following algorithm, described in CIRC-1207:
1. If a single loan has multiple Lost Item Fees/Lost Item Processing Fees, look for the most recent one.
- If the most recent Lost Item Fee has a Payment Status that contains "Cancelled", there is nothing to refund and you should move on.
- If the most recent Lost Item Fee has a Payment Status that does not contain "Cancelled", then continue on.
2. Use the date/time of the most recent Lost Item Fee to locate the corresponding Lost Item Processing Fee (if there is one). The date/time of the Lost Item Processing Fee may be 1 minute earlier or 1 minute later.
- If the corresponding Lost Item Processing Fee has a Payment Status that contains "Cancelled", you will focus only on the Lost Item Fee.
- If the corresponding Lost Item Processing Fee has a Payment Status that does not contain "Cancelled", you will focus on both the Lost Item Fee and the Lost Item Processing Fee.
3. Check the selected Lost Item Fee and Lost Item Processing Fee to see if anything was paid/transferred but not already refunded.
- If the most recent Lost Item Fee has a Payment Status that contains "Cancelled", or there is was never a Lost Item Fee billed, there is nothing to refund and you should move on.
- If the most recent Lost Item Processing Fee has a Payment Status that contains "Cancelled", or there is was never a Lost Item Processing Fee billed, there is nothing to refund and you should move on.

Resolves: [CIRC-1223](https://issues.folio.org/browse/CIRC-1223)

## Approach
- Update logic for refund;
- Create test with the scenario similar to the one described in the ticket;

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.

### CODE REVIEW

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java [hunk 1]
```diff
@@ -203,8 +213,55 @@ private static boolean shouldSchedulePatronNotices(LostItemFeeRefundContext cont
...
+      return accountRepository.findAccountsAndActionsForLoanByQuery(fetchQuery)
+        .thenApply(r -> r.map(accounts -> filterAccountsForRefund(accounts, feeFineTypes)))
+        .thenApply(r -> r.map(context::withAccounts));
+  }
+
+  private Collection<Account> filterAccountsForRefund(Collection<Account> accounts,
+    List<String> feeFineTypes) {
+
+    return getLatestAccount(accounts, feeFineTypes)
+      .filter(this::isAccountValidForRefund)
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java:225]: isAccountEligibleForRefund

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java [hunk 2]
```diff
@@ -203,8 +213,55 @@ private static boolean shouldSchedulePatronNotices(LostItemFeeRefundContext cont
...
+      .map(account -> findRefundableAccounts(account, accounts))
+      .orElse(Collections.emptyList());
+  }
+
+  private List<Account> findRefundableAccounts(Account latestAccount, Collection<Account> accounts) {
+    List<Account> filteredList = new ArrayList<>();
+    filteredList.add(latestAccount);
+    DateTime creationDate = latestAccount.getCreationDate();
+    List<String> feeFineTypeForSearch = LOST_ITEM_FEE_TYPE.equals(latestAccount.getFeeFineType())
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java:234]: List<String> feeFineTypeForSearch = listOf(LOST_ITEM_FEE_TYPE.equals(latestAccount.getFeeFineType()) 
? LOST_ITEM_PROCESSING_FEE_TYPE 
: LOST_ITEM_FEE_TYPE)

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java [hunk 3]
```diff
@@ -203,8 +213,55 @@ private static boolean shouldSchedulePatronNotices(LostItemFeeRefundContext cont
...
+      ? Collections.singletonList(LOST_ITEM_PROCESSING_FEE_TYPE)
+      : Collections.singletonList(LOST_ITEM_FEE_TYPE);
+
+    getLatestAccount(accounts, feeFineTypeForSearch)
+      .filter(this::isAccountValidForRefund)
+      .filter(latestProcessingAccount -> isDifferenceOneMinuteOrLess(creationDate,
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java:240]: secondAccount

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java [hunk 4]
```diff
@@ -203,8 +213,55 @@ private static boolean shouldSchedulePatronNotices(LostItemFeeRefundContext cont
...
+
+  private List<Account> findRefundableAccounts(Account latestAccount, Collection<Account> accounts) {
+    List<Account> filteredList = new ArrayList<>();
+    filteredList.add(latestAccount);
+    DateTime creationDate = latestAccount.getCreationDate();
+    List<String> feeFineTypeForSearch = LOST_ITEM_FEE_TYPE.equals(latestAccount.getFeeFineType())
+      ? Collections.singletonList(LOST_ITEM_PROCESSING_FEE_TYPE)
+      : Collections.singletonList(LOST_ITEM_FEE_TYPE);
+
+    getLatestAccount(accounts, feeFineTypeForSearch)
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java:238]: firstAccount

File: src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java [hunk 5]
```diff
@@ -203,8 +213,55 @@ private static boolean shouldSchedulePatronNotices(LostItemFeeRefundContext cont
...
+
+  private boolean isAccountValidForRefund(Account latestLostItemFeeAccount) {
+    return latestLostItemFeeAccount.getPaymentStatus() != null
+      && !latestLostItemFeeAccount.getPaymentStatus().startsWith(CANCELLED_PAYMENT_STATUS_PREFIX)
+      && latestLostItemFeeAccount.getCreationDate() != null;
+  }
+
+  private boolean isDifferenceOneMinuteOrLess(DateTime latestFeeFineDate, DateTime associatedFeeFineDate) {
+    return secondsBetween(latestFeeFineDate, associatedFeeFineDate)
+      .getSeconds() <= MAX_TIME_INTERVAL_FOR_ASSOCIATED_ACCOUNTS;
```
Comment (roman-barannyk) [inline src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java:255]:
```suggestion
      .getSeconds() <= MAX_TIME_DIFFERENCE_FOR_ASSOCIATED_ACCOUNTS;
```

**Review by alexanderkurash — APPROVED** [review]

**Pull request merged**